### DISCUSSION TITLE
"Challenges" in Developing a Payment Module

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#6120] [General]
TL;DR

**_Note_**: This discussion is meant to be informative-only.  Although I'd like to see a "re-architecture" of an order-total module's interactions when an order is built, I'm not pushing that as a suggestion.

Payment modules have a couple of significant methods called during an order's generation, primarily in the *confirmation* and *process* phases of the checkout process.  The sequencing of the order's generation in those two phases is subtly different.

- During the *confirmation* phase of checkout, the associated `header_php.php` builds the order using this sequence:
  - Instantiate the order-object.
  - Instantiate the active shipping-module; it adds its cost and tax to the order-object.
  - Instantiate the `order_total` class, which loads all the active order-total modules.
  - Run the order-totals'  `collect_posts` and `pre_confirmation_check` methods.
  - Instantiate the active payment-module and run its `update_status` method.
  - Run the active payment-module's `pre_confirmation_check` method.
- During the *process* phase of checkout, a combination of the `checkout_process.php` module and the `checkout_process` page's `header_php.php` build and record the order using the following sequence.  Note that all steps in the sequence are performed ***after*** the session's `payment_attempt` counter is incremented!
  - Instantiate the active payment-module, noting that since the order-object is not yet instantiated the payment-module's `update_status` method isn't run.
  - Instantiate the order-object.
  - Instantiate the active shipping-module; it adds its cost and tax to the order-object.
  - Instantiate the `order_total` class, which loads all the active order-total modules.
  - Run the order-totals' `pre_confirmation_check` method if the active payment-module's 'code' *doesn't start with* `paypal`.
    - Why this distinction, I'm unclear, but that's another discussion.
  - Run the order-totals' `process` method.
  - Run the active payment-module's `before_process` method.
  - Create and record the order's base information; at this point the order's `orders_id` is known.
  - Run the active payment-module's `after_order_create` method.
  - Create the `orders_products` table entries for the ordered products.
  - Send the order-completion email.
  - Run the active payment-module's `after_process` method.  Note that this is the only step not run by the `checkout_process.php` module.

***Note***: The `order_total` class' `pre_confirmation_check` call actually runs each order-total's `process` method, saving and restoring the order-object's `info` element during that processing.

The payment-module I've been working on supports an order's amount-breakdown.  Supplying that breakdown provides the store-owner some additional protection for fraudulent payments.  As such, I've chosen to use the payment-module's `pre_confirmation_check` to gather that breakdown and do an initial submission to the payment-processor during the confirmation phase of checkout so that any rudimentary issues with the payment's creation doesn't "count against" the customer.

What I discovered is that the order-object isn't well-formed at the point where the payment's `pre_confirmation_check` is called during the *confirmation* phase.  Here's an example order that I'll use for further explanation.

1. I've got the `loworder_fee` module enabled with a $5.00 fee taxable with 'Taxable Goods' for orders less than $50.00.
2. Using the default US/FL tax-rate of 7%.
3. Using flat-rate shipping, with a $5.00 fee taxable with 'Taxable Goods'.
4. Using `index.php?main_page=product_info&cPath=54_55&products_id=127` as an example, a $15.00 product.
5. Customer's order is shipping to US/FL.

On entry to the payment's `pre_confirmation_check`, here's what the pertinent amount-related values are in `$order->info`:

```
    "currency": "USD",
    "currency_value": "1.000000",
    "coupon_code": "",
    "shipping_method": "Flat Rate (Best Way)",
    "shipping_module_code": "flat_flat",
    "shipping_cost": "5.00",
    "subtotal": 15,
    "shipping_tax": 0,          <== No shipping tax, should be 0.35
    "shipping_tax_rate": null,  <== Shipping tax-rate is actually 7.0
    "tax": 1.050000000000001,   <== Tax value is based only on products (15.00 * 0.07)
    "total": 21.05,             <== Total doesn't include low-order fee, fee's taxes or shipping tax (should be 26.75)
    "tax_groups": {
        "FL TAX 7.0%": 1.050000000000001
    },
```

That behavior was the basis of [this](https://github.com/zencart/zencart/pull/6090) PR on the `order_total` class.  With those notifications, the payment-module includes an auto-loaded observer that can grab the totals' calculations and successfully provide the amount-breakdown required by the payment-processor.

The order-object is well-formed when the payment-module's `before_process` method is called, but I'm still using the results from that observer-class to more easily determine the order's amount-breakdown:

```
    "currency": "USD",
    "currency_value": "1.000000",
    "coupon_code": "",
    "shipping_method": "Flat Rate (Best Way)",
    "shipping_module_code": "flat_flat",
    "shipping_cost": "5.00",
    "subtotal": 15,
    "shipping_tax": 0.3500000000000003,
    "shipping_tax_rate": 7.000000000000006,
    "tax": 1.7500000000000016,
    "total": 26.750000000000004,
    "tax_groups": {
        "FL TAX 7.0%": 1.7500000000000016
    },
```
